### PR TITLE
Abdul/nft metadata json

### PR DESCRIPTION
**Pull request by imabdulbasit**
Metadata from URI is parsed and inserted into database
tables created: 
metadata_jsons
attributes
files
metadata_collections

attributes, files and metadata_collections tables may have NULL values for attributes other than metadata_address and  Diesel doesn't allow tables without primary key so had to insert an id variable as primary key.

### CODE REVIEW

**Review by kespinola — DISMISSED** [review]
This looks good to me. Only items that come up are indexes and we arent capturing the change states of metadata json since the uri field of nft is mutable. Items to think about in the future.

**Comment by imabdulbasit** [comment]
> This looks good to me. Only items that come up are indexes and we arent capturing the change states of metadata json since the uri field of nft is mutable. Items to think about in the future.

So one metadata_address has one fingerprint, a check can added if fingerprint is different for a metadata_address

**Comment by kespinola** [comment]
Notes from initial testing:

```
metaplex-indexer=# SELECT COUNT(*) FROM metadata_jsons;;
 count
-------
  4594
(1 row)

metaplex-indexer=# SELECT COUNT(*) FROM metadatas;;
 count
-------
 10580
(1 row)
```

In a fresh run about half metadata have an associated json.


```
[2022-01-24T15:05:37Z DEBUG reqwest::async_impl::client] response '200 OK' for https://bqrzzb3nqwavyhn7c6cpvoloojoh3cvrf52emv7b5ldzuuiwazaa.arweave.net/DCOch22FgVwdvxeE-rluclx9irEvdEZX4erHmlEWBkA/
[2022-01-24T15:08:32Z ERROR metaplex_indexer::entry] Job MetadataUri(5btasmbUigzK7wstC1UpasrLHEVC6cYPxwo44KupXxCr, "https://arweave.net/R1U8tgnWrahVf2KpaISSGWWxbjgHRirQKyqmMn6uPG0") failed: Metadata JSON request failed

    Caused by:
        0: error sending request for url (https://arweave.net/R1U8tgnWrahVf2KpaISSGWWxbjgHRirQKyqmMn6uPG0): error trying to connect: connection closed via error
        1: error trying to connect: connection closed via error
        2: connection closed via error
```

Error with connection timeout to arweave.

Don't believe I had our CDNs in place trying again with those.

File: crates/core/src/lib.rs [hunk 1]
```diff
@@ -58,7 +58,7 @@ fn dotenv(name: impl AsRef<Path>) -> Result<Option<PathBuf>, dotenv::Error> {
 #[derive(Debug, Clone, Copy, clap::Parser)]
 pub struct ServerOpts {
     /// The port to listen on
-    #[clap(short, long, default_value_t = 3000, env)]
+    #[clap(short, long, default_value_t = 3000, env = "PORT")]
```
Comment (ray-kast) [inline crates/core/src/lib.rs:61]:
```suggestion
    #[clap(short, long, default_value_t = 3000, env)]
```
The `= "PORT"` should be implicit, was this added back for a reason?

File: crates/core/src/db/models.rs [hunk 2]
```diff
@@ -226,3 +211,81 @@ pub struct ListingsTripleJoinRow {
...
+    pub metadata_address: Cow<'a, str>,
+    /// Attribute name
+    pub name: Option<String>,
+    /// Attribute value
+    pub value: Option<String>,
+    /// Attribute trait type
+    pub trait_type: Option<String>,
+}
+
+/// A row in 'metadata_collection' table
```
Comment (ray-kast) [inline crates/core/src/db/models.rs:265]:
```suggestion
/// A row in the `metadata_collections` table
```

File: crates/indexer/Cargo.toml [hunk 3]
```diff
@@ -14,6 +14,9 @@ keywords = ["solana", "metaplex", "holaplex", "web3"]
 categories = ["cryptography::cryptocurrencies", "web-programming"]
 
 [dependencies]
+base64 = "0.13.0"
+cid = "0.7.0"
+clap = "3.0.0-beta.5"
```
Comment (ray-kast) [inline crates/indexer/Cargo.toml:19]:
```suggestion
```
This dependency was removed in another branch.

File: crates/indexer/src/bits/metadata_uri.rs [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+    animation_url: Option<String>,
+    collection: Option<Collection>,
+    attributes: Option<Vec<Attribute>>,
+    external_url: Option<String>,
+    properties: Option<Property>,
+    #[serde(flatten)]
+    extra: HashMap<String, serde_json::Value>,
+}
+
+async fn process_async<'a>(
```
Comment (ray-kast) [inline crates/indexer/src/bits/metadata_uri.rs:156]: This function has some linter warnings that should be addressed.

File: crates/indexer/src/bits/token_account.rs [hunk 5]
```diff
@@ -1,53 +0,0 @@
...
-    insert_into(token_accounts::table)
-        .values(&values)
-        .on_conflict(token_accounts::address)
-        .do_update()
-        .set(&values)
-        .execute(&db)
-        .context("failed to insert token account")?;
-
-    Ok(())
-}
```
Comment (ray-kast) [inline crates/indexer/src/bits/token_account.rs:53]: Why is this removal in this PR?

**Review by ray-kast — CHANGES_REQUESTED** [review]
I think the merge from dev generated some broken diffs, the entire PR likely needs to be rebased onto the new dev.

**Comment by kespinola** [comment]
having some git problems so moving over to the fork of this pr directly into dev. commits retained but was hoping to merge here first.

https://github.com/holaplex/metaplex-indexer/pull/79

**Pull request closed without merging**